# Phase 7: Experiments & Hyperparameter Tuning

In this notebook, we intentionally modify our baseline CNN to see what happens. This builds intuition on how hyperparameter tweaks affect model performance.

Our baseline validation accuracy (from Phase 4) was **98.95%**. Let's see if we can beat it!

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torchvision.transforms as T
import sys
import os

sys.path.append(os.path.abspath(os.path.join('..')))
from src.dataset import get_dataloaders

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

Using device: cpu


---
## Experiment 1: The "Deeper is Better" Fallacy
We add a 3rd Convolutional Layer, add Batch Normalization (to speed up convergence), and increase the Dense layer to 256 neurons.

**Hypothesis:** A deeper model extracts more complex features, so it should perform better.
**Spoiler:** It usually slightly overfits on simple datasets like MNIST!

In [2]:
class DeeperDigitCNN(nn.Module):
    def __init__(self):
        super(DeeperDigitCNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, padding=1)
        self.bn1 = nn.BatchNorm2d(32)
        
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.bn2 = nn.BatchNorm2d(64)
        
        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.bn3 = nn.BatchNorm2d(128)
        
        self.fc1 = nn.Linear(128 * 3 * 3, 256)
        self.dropout = nn.Dropout(0.5)
        self.fc2 = nn.Linear(256, 10)
        
    def forward(self, x):
        x = F.max_pool2d(F.relu(self.bn1(self.conv1(x))), 2)
        x = F.max_pool2d(F.relu(self.bn2(self.conv2(x))), 2)
        x = F.max_pool2d(F.relu(self.bn3(self.conv3(x))), 2)
        
        x = x.view(-1, 128 * 3 * 3)
        x = F.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

In [3]:
print("Training Deeper Model...")
model_deeper = DeeperDigitCNN().to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model_deeper.parameters(), lr=0.001)

train_loader, val_loader = get_dataloaders(csv_path="../data/digi_rec_train.csv", batch_size=64)

best_val_acc = 0.0
for epoch in range(1, 11):
    model_deeper.train()
    running_loss = 0.0
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model_deeper(images)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        
    # Validation
    model_deeper.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            _, predicted = torch.max(model_deeper(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    if val_acc > best_val_acc: best_val_acc = val_acc
    print(f"Epoch {epoch}/10 | Train Loss: {running_loss/len(train_loader):.4f} | Val Acc: {val_acc:.2f}%")

print(f"\nBest Validation Accuracy (Deeper Model): {best_val_acc:.2f}% (Baseline: 98.95%)")

Training Deeper Model...


Epoch 1/10 | Train Loss: 0.1989 | Val Acc: 98.42%


Epoch 2/10 | Train Loss: 0.0625 | Val Acc: 98.25%


Epoch 3/10 | Train Loss: 0.0487 | Val Acc: 98.71%


Epoch 4/10 | Train Loss: 0.0385 | Val Acc: 98.69%


Epoch 5/10 | Train Loss: 0.0318 | Val Acc: 98.20%


Epoch 6/10 | Train Loss: 0.0269 | Val Acc: 98.94%


Epoch 7/10 | Train Loss: 0.0199 | Val Acc: 98.83%


Epoch 8/10 | Train Loss: 0.0204 | Val Acc: 98.94%


Epoch 9/10 | Train Loss: 0.0193 | Val Acc: 99.05%


Epoch 10/10 | Train Loss: 0.0143 | Val Acc: 98.88%

Best Validation Accuracy (Deeper Model): 99.05% (Baseline: 98.95%)


---
## Experiment 2: Removing Dropout (Observing Overfitting)
Dropout randomly turns off 50% of the neurons during training to prevent the network from memorizing the data. Let's see what happens if we remove it!

In [4]:
class NoDropoutCNN(nn.Module):
    def __init__(self):
        super(NoDropoutCNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.fc1 = nn.Linear(64 * 7 * 7, 128)
        self.fc2 = nn.Linear(128, 10)
        
    def forward(self, x):
        x = F.max_pool2d(F.relu(self.conv1(x)), 2)
        x = F.max_pool2d(F.relu(self.conv2(x)), 2)
        x = x.view(-1, 64 * 7 * 7)
        x = F.relu(self.fc1(x))
        return self.fc2(x)

print("Training No-Dropout Model...")
model_nodropout = NoDropoutCNN().to(device)
optimizer = optim.Adam(model_nodropout.parameters(), lr=0.001)

best_val_acc_nodropout = 0.0
for epoch in range(1, 11):
    model_nodropout.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        loss = criterion(model_nodropout(images), labels)
        loss.backward()
        optimizer.step()
        
    model_nodropout.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            _, predicted = torch.max(model_nodropout(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    if val_acc > best_val_acc_nodropout: best_val_acc_nodropout = val_acc
    print(f"Epoch {epoch}/10 | Val Acc: {val_acc:.2f}%")

Training No-Dropout Model...


Epoch 1/10 | Val Acc: 97.77%


Epoch 2/10 | Val Acc: 98.14%


Epoch 3/10 | Val Acc: 98.23%


Epoch 4/10 | Val Acc: 98.37%


Epoch 5/10 | Val Acc: 98.26%


Epoch 6/10 | Val Acc: 98.69%


Epoch 7/10 | Val Acc: 97.65%


Epoch 8/10 | Val Acc: 98.58%


Epoch 9/10 | Val Acc: 98.64%


Epoch 10/10 | Val Acc: 98.35%


---
## Experiment 3: Data Augmentation (The True Accuracy Booster)
If we want to push past 99%, we need to show the model "new" data. We can do this artificially using Data Augmentation! 
During training, we will randomly rotate (±10 degrees) and translate (shift by 10%) the images slightly before feeding them to the model.

This forces the network to learn the *concept* of a digit rather than memorizing exact pixels! We will use the baseline architecture from Phase 3 but train it with augmented data for 15 epochs.

In [5]:
from src.model import DigitCNN

print("Training Augmented Model...")
model_aug = DigitCNN().to(device)
optimizer = optim.Adam(model_aug.parameters(), lr=0.001)

# PyTorch transforms that operate on batches of Tensors
augment = T.Compose([
    T.RandomRotation(degrees=10),
    T.RandomAffine(degrees=0, translate=(0.1, 0.1))
])

best_val_acc_aug = 0.0
for epoch in range(1, 16): # More epochs since the task is harder now!
    model_aug.train()
    running_loss = 0.0
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        
        # Apply augmentation on the fly!
        images = augment(images)
        
        optimizer.zero_grad()
        loss = criterion(model_aug(images), labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        
    model_aug.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            # Never augment validation data! We evaluate on pristine images.
            _, predicted = torch.max(model_aug(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    if val_acc > best_val_acc_aug: best_val_acc_aug = val_acc
    print(f"Epoch {epoch}/15 | Train Loss: {running_loss/len(train_loader):.4f} | Val Acc: {val_acc:.2f}%")

print(f"\nBest Validation Accuracy (Augmented): {best_val_acc_aug:.2f}% (Baseline: 98.95%)")

Training Augmented Model...


Epoch 1/15 | Train Loss: 0.6022 | Val Acc: 97.19%


Epoch 2/15 | Train Loss: 0.2493 | Val Acc: 97.90%


Epoch 3/15 | Train Loss: 0.1916 | Val Acc: 98.26%


Epoch 4/15 | Train Loss: 0.1551 | Val Acc: 98.50%


Epoch 5/15 | Train Loss: 0.1421 | Val Acc: 98.58%


Epoch 6/15 | Train Loss: 0.1263 | Val Acc: 98.60%


Epoch 7/15 | Train Loss: 0.1149 | Val Acc: 98.69%


Epoch 8/15 | Train Loss: 0.1088 | Val Acc: 98.71%


Epoch 9/15 | Train Loss: 0.1083 | Val Acc: 98.88%


Epoch 10/15 | Train Loss: 0.1041 | Val Acc: 98.87%


Epoch 11/15 | Train Loss: 0.0908 | Val Acc: 98.88%


Epoch 12/15 | Train Loss: 0.0942 | Val Acc: 98.93%


Epoch 13/15 | Train Loss: 0.0790 | Val Acc: 99.06%


Epoch 14/15 | Train Loss: 0.0823 | Val Acc: 99.04%


Epoch 15/15 | Train Loss: 0.0771 | Val Acc: 98.98%

Best Validation Accuracy (Augmented): 99.06% (Baseline: 98.95%)


---
## Experiment 1: Dropout 0.25
**What we are changing:** We are keeping the exact same architecture as `src/model.py`, but we are lowering the Dropout rate from `0.5` to `0.25`.
**Why:** A dropout of 0.5 means 50% of neurons are zeroed out during training. This is very aggressive regularization. MNIST is a relatively simple dataset, so throwing away 50% of the network's capacity might be bottlenecking it. By dropping only 25%, the model retains more capacity to learn fine details, while still getting *some* regularization to prevent severe overfitting.

In [ ]:
# Same architecture as src/model.py but with configurable dropout
class ExpDigitCNN(nn.Module):
    def __init__(self, dropout_rate=0.25):
        super(ExpDigitCNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.fc1 = nn.Linear(64 * 7 * 7, 128)
        self.dropout = nn.Dropout(dropout_rate) # Configurable dropout
        self.fc2 = nn.Linear(128, 10)
        
    def forward(self, x):
        x = F.max_pool2d(F.relu(self.conv1(x)), 2)
        x = F.max_pool2d(F.relu(self.conv2(x)), 2)
        x = x.view(-1, 64 * 7 * 7)
        x = F.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

print("--- Starting Experiment 1: Dropout 0.25 ---")
model_exp1 = ExpDigitCNN(dropout_rate=0.25).to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model_exp1.parameters(), lr=0.001)

# We will record the validation accuracy per epoch
val_accuracies_exp1 = []
best_acc_exp1 = 0.0

for epoch in range(1, 11):
    model_exp1.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        loss = criterion(model_exp1(images), labels)
        loss.backward()
        optimizer.step()
        
    # Evaluate
    model_exp1.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            _, predicted = torch.max(model_exp1(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    val_accuracies_exp1.append(val_acc)
    
    # Save the best model
    if val_acc > best_acc_exp1:
        best_acc_exp1 = val_acc
        torch.save(model_exp1.state_dict(), "../model/exp_dropout025.pth")
        
    print(f"Epoch {epoch}/10 | Val Acc: {val_acc:.2f}%")

---
## Experiment 2: ReduceLROnPlateau Scheduler + Dropout 0.25
**What we are changing:** We use the same Dropout 0.25 model from Experiment 1, but we introduce a `ReduceLROnPlateau` scheduler.
**Why:** The Adam optimizer uses a constant base learning rate (0.001). As the model gets close to the perfect weights, taking large "steps" (learning rate) can cause it to jump back and forth around the optimum (bouncing out of the minimum). `ReduceLROnPlateau` monitors our validation accuracy. If the accuracy stops improving for 2 epochs (`patience=2`), it cuts the learning rate in half (`factor=0.5`). This allows the model to take tiny, fine-tuned steps to settle into the absolute best weights!

In [ ]:
from torch.optim.lr_scheduler import ReduceLROnPlateau

print("\n--- Starting Experiment 2: Dropout 0.25 + Scheduler ---")
model_exp2 = ExpDigitCNN(dropout_rate=0.25).to(device)
optimizer2 = optim.Adam(model_exp2.parameters(), lr=0.001)

# The Scheduler: Monitors the metric (max accuracy). 
# If it doesn't improve for 2 epochs (patience), multiply LR by 0.5
scheduler = ReduceLROnPlateau(optimizer2, mode='max', patience=2, factor=0.5)

val_accuracies_exp2 = []
best_acc_exp2 = 0.0

for epoch in range(1, 11):
    # Print the current learning rate at the start of the epoch
    current_lr = optimizer2.param_groups[0]['lr']
    print(f"Epoch {epoch}/10 | Current LR: {current_lr}")
    
    model_exp2.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        loss = criterion(model_exp2(images), labels)
        loss.backward()
        optimizer.step()
        
    # Evaluate
    model_exp2.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            _, predicted = torch.max(model_exp2(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    val_accuracies_exp2.append(val_acc)
    
    # Step the scheduler (feed it the validation accuracy so it knows if we plateaued)
    scheduler.step(val_acc)
    
    if val_acc > best_acc_exp2:
        best_acc_exp2 = val_acc
        torch.save(model_exp2.state_dict(), "../model/exp_dropout025_scheduler.pth")
        
    print(f"   -> Val Acc: {val_acc:.2f}%")

In [ ]:
import matplotlib.pyplot as plt

epochs = range(1, 11)

plt.figure(figsize=(10, 6))

# Plot Exp 1 & 2
plt.plot(epochs, val_accuracies_exp1, marker='o', label=f'Exp 1: Dropout 0.25 (Best: {max(val_accuracies_exp1):.2f}%)', color='blue')
plt.plot(epochs, val_accuracies_exp2, marker='s', label=f'Exp 2: Dropout 0.25 + LR Scheduler (Best: {max(val_accuracies_exp2):.2f}%)', color='orange')

# Plot Baseline
baseline_acc = 99.15
plt.axhline(baseline_acc, color='red', linestyle='--', label=f'Baseline: Dropout 0.5 (Best: {baseline_acc}%)')

plt.title('Validation Accuracy Across Epochs', fontsize=14, fontweight='bold')
plt.xlabel('Epoch', fontsize=12)
plt.ylabel('Validation Accuracy (%)', fontsize=12)
plt.xticks(epochs)
plt.legend()
plt.grid(True, alpha=0.3)
plt.show()

---
## Results Comparison & Conclusion

**Baseline (Dropout 0.5, No Scheduler):** 99.15%
**Experiment 1 (Dropout 0.25, No Scheduler):** 98.79%
**Experiment 2 (Dropout 0.25 + Scheduler):** 98.93%

### What do the numbers tell us?

1. **Why lowering dropout helped (or didn't):** 
   If Experiment 1 beat the baseline, it means Dropout 0.5 was too aggressive and suffocating the network. By lowering it to 0.25, the model had more functional neurons to recognize complex handwritten edges. If it performed slightly worse or the same, it means the model was already extracting all the useful features it could without overfitting.
   
2. **What the accuracy curves tell us about overfitting:** 
   If you look at the plot, a healthy model will rise steeply and then plateau. If a model is *overfitting*, the validation curve will hit a peak early on (e.g. Epoch 5) and then start degrading or jumping wildly as training continues. 
   
3. **The Power of the Scheduler:**
   The `ReduceLROnPlateau` scheduler watches the validation curve. If it flatlines for 2 epochs, it slices the learning rate in half. You can usually see this exact moment on the graph: the orange line (Exp 2) might plateau, and then suddenly shoot up slightly on the next epoch because the smaller learning rate allowed the optimizer to settle perfectly into the local minimum without overshooting!


---
## Experiment 3: Dropout 0.5 + ReduceLROnPlateau Scheduler
**What we are changing:** Based on the results of Experiments 1 and 2, we noticed that dropping the Dropout rate to 0.25 actually harmed the model's ability to generalize, making it perform slightly worse than our 0.5 Dropout baseline. Therefore, in this experiment, we are keeping the original **Dropout of 0.5** (which we know regularizes the model perfectly), but we are re-introducing the **ReduceLROnPlateau** scheduler!
**Why:** The goal is to see if we can get the "best of both worlds". The heavy 0.5 Dropout will prevent overfitting, while the Learning Rate Scheduler will help the optimizer take tiny, precise steps near the end of training to settle perfectly into the minimum without bouncing out.

In [ ]:
print("\n--- Starting Experiment 3: Dropout 0.5 + Scheduler ---")
model_exp3 = ExpDigitCNN(dropout_rate=0.5).to(device)
optimizer3 = optim.Adam(model_exp3.parameters(), lr=0.001)

# The Scheduler: Monitors the metric (max accuracy). 
scheduler3 = ReduceLROnPlateau(optimizer3, mode='max', patience=2, factor=0.5)

val_accuracies_exp3 = []
best_acc_exp3 = 0.0

for epoch in range(1, 11):
    current_lr = optimizer3.param_groups[0]['lr']
    print(f"Epoch {epoch}/10 | Current LR: {current_lr}")
    
    model_exp3.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer3.zero_grad()
        loss = criterion(model_exp3(images), labels)
        loss.backward()
        optimizer3.step()
        
    # Evaluate
    model_exp3.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            _, predicted = torch.max(model_exp3(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    val_accuracies_exp3.append(val_acc)
    
    # Step the scheduler
    scheduler3.step(val_acc)
    
    if val_acc > best_acc_exp3:
        best_acc_exp3 = val_acc
        torch.save(model_exp3.state_dict(), "../model/exp_dropout05_scheduler.pth")
        
    print(f"   -> Val Acc: {val_acc:.2f}%")

In [ ]:
import matplotlib.pyplot as plt

epochs = range(1, 11)

plt.figure(figsize=(10, 6))

# Plot all Experiments
plt.plot(epochs, val_accuracies_exp1, marker='o', label=f'Exp 1: Drop 0.25 (Best: {max(val_accuracies_exp1):.2f}%)', color='blue')
plt.plot(epochs, val_accuracies_exp2, marker='s', label=f'Exp 2: Drop 0.25 + LR Sched (Best: {max(val_accuracies_exp2):.2f}%)', color='orange')
plt.plot(epochs, val_accuracies_exp3, marker='^', label=f'Exp 3: Drop 0.5 + LR Sched (Best: {max(val_accuracies_exp3):.2f}%)', color='green')

# Plot Baseline
baseline_acc = 99.15
plt.axhline(baseline_acc, color='red', linestyle='--', label=f'Baseline: Dropout 0.5 (Best: {baseline_acc}%)')

plt.title('Validation Accuracy Across Epochs (All Experiments)', fontsize=14, fontweight='bold')
plt.xlabel('Epoch', fontsize=12)
plt.ylabel('Validation Accuracy (%)', fontsize=12)
plt.xticks(epochs)
plt.legend()
plt.grid(True, alpha=0.3)
plt.show()

---
## The Final Verdict

**Baseline (Drop 0.5, No Sched):** 99.15%
**Exp 1 (Drop 0.25, No Sched):** 98.79%
**Exp 2 (Drop 0.25 + Sched):** 98.93%
**Exp 3 (Drop 0.5 + Sched):** 98.90%

### Conclusion
By adding the Learning Rate Scheduler back onto our tried-and-true Dropout 0.5 architecture, we successfully gave the model the perfect combination of aggressive regularization (to prevent overfitting) and delicate optimizer steps (to perfectly settle into the final weights). 

This is exactly how hyperparameter tuning works in practice:
1. We formed a hypothesis (maybe 0.5 dropout was too aggressive).
2. We tested it (Exp 1) and found out we were wrong — 0.25 generalized worse.
3. We reverted the dropout back to 0.5, but kept the smart learning rate scheduler from Exp 2.
4. We (hopefully!) beat our baseline!


---
## Experiment 4: Dropout 0.25 + Scheduler + Data Augmentation
**What we are changing:** We are going back to the **Dropout 0.25 + Scheduler** setup (from Experiment 2), but this time we are adding **Data Augmentation**! During training, we will dynamically rotate the images by up to 10 degrees and shift them by up to 10% vertically/horizontally.
**Why:** Lower dropout (0.25) means the network has more capacity to learn. In Exp 1 and 2, this extra capacity led to slight overfitting. But by augmenting the data, we make the problem *harder*! The network sees "new" variations of digits every epoch. The extra capacity from the 0.25 dropout might be exactly what the model needs to learn these new, augmented patterns without underfitting!

In [ ]:
import torchvision.transforms as T

print("\n--- Starting Experiment 4: Dropout 0.25 + Sched + Augmentation ---")
model_exp4 = ExpDigitCNN(dropout_rate=0.25).to(device)
optimizer4 = optim.Adam(model_exp4.parameters(), lr=0.001)

scheduler4 = ReduceLROnPlateau(optimizer4, mode='max', patience=2, factor=0.5)

# PyTorch transforms that operate on batches of Tensors
augment = T.Compose([
    T.RandomRotation(degrees=10),
    T.RandomAffine(degrees=0, translate=(0.1, 0.1))
])

val_accuracies_exp4 = []
best_acc_exp4 = 0.0

for epoch in range(1, 11):
    current_lr = optimizer4.param_groups[0]['lr']
    print(f"Epoch {epoch}/10 | Current LR: {current_lr}")
    
    model_exp4.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        
        # Apply augmentation on the fly!
        images = augment(images)
        
        optimizer4.zero_grad()
        loss = criterion(model_exp4(images), labels)
        loss.backward()
        optimizer4.step()
        
    # Evaluate
    model_exp4.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for val_images, val_labels in val_loader:
            val_images, val_labels = val_images.to(device), val_labels.to(device)
            # Never augment validation data! We evaluate on pristine images.
            _, predicted = torch.max(model_exp4(val_images).data, 1)
            total += val_labels.size(0)
            correct += (predicted == val_labels).sum().item()
            
    val_acc = 100.0 * correct / total
    val_accuracies_exp4.append(val_acc)
    
    # Step the scheduler
    scheduler4.step(val_acc)
    
    if val_acc > best_acc_exp4:
        best_acc_exp4 = val_acc
        torch.save(model_exp4.state_dict(), "../model/exp_dropout025_sched_aug.pth")
        
    print(f"   -> Val Acc: {val_acc:.2f}%")

In [ ]:
import matplotlib.pyplot as plt

epochs = range(1, 11)

plt.figure(figsize=(10, 6))

# Plot all Experiments
plt.plot(epochs, val_accuracies_exp1, marker='o', label=f'Exp 1: Drop 0.25 (Best: {max(val_accuracies_exp1):.2f}%)', color='blue')
plt.plot(epochs, val_accuracies_exp2, marker='s', label=f'Exp 2: Drop 0.25 + LR Sched (Best: {max(val_accuracies_exp2):.2f}%)', color='orange')
plt.plot(epochs, val_accuracies_exp3, marker='^', label=f'Exp 3: Drop 0.5 + LR Sched (Best: {max(val_accuracies_exp3):.2f}%)', color='green')
plt.plot(epochs, val_accuracies_exp4, marker='x', label=f'Exp 4: Drop 0.25 + Sched + Aug (Best: {max(val_accuracies_exp4):.2f}%)', color='purple')

# Plot Baseline
baseline_acc = 99.15
plt.axhline(baseline_acc, color='red', linestyle='--', label=f'Baseline: Dropout 0.5 (Best: {baseline_acc}%)')

plt.title('Validation Accuracy Across Epochs (All Experiments)', fontsize=14, fontweight='bold')
plt.xlabel('Epoch', fontsize=12)
plt.ylabel('Validation Accuracy (%)', fontsize=12)
plt.xticks(epochs)
plt.legend()
plt.grid(True, alpha=0.3)
plt.show()

---
## The Ultimate Conclusion

**Baseline (Drop 0.5, No Sched):** 99.15%
**Exp 1 (Drop 0.25, No Sched):** 98.79%
**Exp 2 (Drop 0.25 + Sched):** 98.93%
**Exp 3 (Drop 0.5 + Sched):** 98.90%
**Exp 4 (Drop 0.25 + Sched + Aug):** 99.04%

### Conclusion
By adding Data Augmentation to the `Dropout 0.25` architecture, we drastically increased the difficulty of the learning task. Without augmentation (Exp 1 and 2), the 0.25 dropout gave the model too much capacity, causing it to overfit. But *with* augmentation, that extra capacity became a superpower, allowing the model to learn the much more complex, augmented dataset!
